### ISSUE TITLE
Thread Safe DependencyProperty Setting/Getting

### ISSUE THREAD

**Issue by legistek** [dotnet/wpf#168]
I'm gonna say at the outset I don't pretend to know all the reasons why setting  `DependencyProperty`s is not threadsafe (in other words, they always must be set on the UI thread), but I do know that it can be a gigantic pain to build responsive multithreaded UIs while dealing with this issue .This is particularly true in the MVVM pattern where we have `DependencyProperty`'s data bound to view model properties, and we would like to change those view model properties in worker threads.

I do realize one major problem is that you can't have properties that might affect layout / measure being changed by other threads during the layout pass, and using thread locking to prevent this would mean that worker threads could get blocked by UI updates, which tends to defeat the purpose of multithreading. 

One possibility, however, might be to allow the `SetProperty` (or more frequently binding source property updates) to succeed when triggered in a worker thread, but internally to use the dispatcher to handle the property change. _Basically do transparently what we now have to do manually._ Until the dispatcher gets to the actual change there would be a slight disconnect between what the viewmodel thinks the value is and what the UI thinks it is, but again we have to deal with this now when using the dispatcher (i.e. we don't know when the effect of property changes will actually propagate to the UI if they occur on a worker thread). 

I could easily be overlooking other difficulties, as well as other possible solutions. Either way I didn't see this among any of the issues opened since the open source announcement and thought I'd raise it to for discussion.

**Comment by dotMorten** [comment]
Wouldn't this be a breaking behavior though? My DP Changed Events and DependencyProeprtyChanged event handlers currently does not expect to get triggered outside the UI thread, so this could break me. For such a change to happen, I think a DP declaration should opt-in to allow multi-threaded usage.

However having said that, I think this is more of a problem at the binding-engine level than at the property level. Currently we have to INPC on the UI thread to avoid binding crashes issues, but that requires me to have the UI Thread bleed into my models, and break the separation of concerns. Add to that that you could actually have multiple UI Threads, so which one do you raise on?

Instead the binding engine is responsible for forwarding the value to the UI Control, and it knows the UI Thread it's working on, so why can't that do the property thread switch and set the DP on the UI thread? I think solving it at the binding level would be much less of a breaking behavior.

**Comment by legistek** [comment]
I don't believe it would be a breaking change actually, at least not the way I'm picturing it. When the `DependencyObject` set methods (`SetValue`, etc.) check for thread agreement and see that the call is on the wrong thread, instead of throwing an exception, they could just re-call `SetValue` a second time, but this time via the `Dispatcher`. The `DependencyProeprtyChanged` event handlers, and the like, would still execute on the UI thread then. It's logically equivalent to what we have to do now.

> Currently we have to INPC on the UI thread to avoid binding crashes issues, but that requires me to have the UI Thread bleed into my models, and break the separation of concerns. Add to that that you could actually have multiple UI Threads, so which one do you raise on?

Well said! Exactly the same issue I have.

**Comment by stevenbrix** [comment]
I think this is a fantastic discussion to have! I do agree this is something we should look into improving, altough the likelihood of breaking behavior means we probably won't get to it for the .NET Core 3 release. We should continue having discussions so we can bring an improvement like like to WPF post .NET Core 3 release.

**Comment by weltkante** [comment]
Would SetValue on the wrong thread block until the set is done or just schedule an update and return?

The former has a high risk of deadlocking because SetValue will inherit and block inside whatever locks the caller currently has taken. The latter makes updates of multiple properties highly fragile because it suddenly becomes random which SetValue wins if multiple threads adjust multiple properties (and not just a single property). Alltogether its also more costly because you need a thread switch for every single property assignment instead of switching once and updating all properties you want to update.

I like the way you currently have to make thread switches explicit because it ensures that you know what you do. Doing it implicitely will either introduce deadlocks or race conditions when you get it wrong. Since threading is hard the most likely scenario is to get it wrong. I would hate having to debug side effects of implicit thread changes.

**Comment by legistek** [comment]
> Would SetValue on the wrong thread block until the set is done or just schedule an update and return?

No blocking; my suggestion at the moment would be to schedule it via the `DependencyObject`'s `Dispatcher` - which is exactly how we have to work around this now - automatically, without throwing an exception.

It doesn't seem like it would be a breaking change because all existing code should make sure setting a DP value or firing an INPC event on the wrong thread never happens. I suppose it's theoretically possible code exists out there that is counting on an exception being thrown because of a DP change on the wrong thread but that seems pretty unlikely.

**Comment by weltkante** [comment]
If you want to set multiple properties you don't do a BeginInvoke for every property you change. At least thats not how I do it. I do a BeginInvoke once and change all properties in bulk. That ensures that my update logic has a consistent view of the object values and nothing else can interleave between updating two properties.

I'm not saying the change would break existing code, I'm saying it would make developing correct code much harder. If I accidently set a property from the wrong thread then I'm now open for a whole bunch of bugs. Instead of pointing me at my mistake via exceptions it just looks like it works, but in production (i.e. outside of debug builds and debuggers, customers with faster/slower machines, etc.) the timing will be totally different so a different interleaving of property changes can happen than what I see in debugging.

Say a background thread wants to update DataContext and Style because it has finished some background loading. So it "usually" works, the DataContext change invalidates and the Style change was scheduled quickly enough so that it runs before the layout. In production the timing could be different and there could be a higher chance that a layout pass will run right after scheduling the DataContext change. Now the layout pass tries to apply a Style to an outdated DataContext, in the best case you just see flicker when the DataContext is changed later, in the worst case you get random crashes in production which you can't make sense of. In general all bets are off because its a genuine race condition.

A completely different problem is GetValue. Do you still want to throw on that, so that you can Set but not Get? Or do you want to do a blocking thread switch to read and return the value? If you somehow manage to make GetValue threadsafe you will be in hell because you can never trust the value you received. There may be outstanding scheduled changes from SetValue which are still in flight. The UI thread may have been in the middle of a layout or binding pass and already have updated some properties but not others.

**Comment by legistek** [comment]
All good points.  Maybe whatever it is should be opt-in only. Give `DependencyObject` an inheritable property like `AllowThreadMismatch` that defaults to `false` for example.

Also maybe it would make sense to do this only in the context of binding source property changes. In other words when handling `PropertyChanged` from INPC. Good view model business logic should be agnostic to UI threading issues anyway. And restricting it to this context only would avoid the need to mess with `GetValue`.

**Comment by lotcz** [comment]
I would love such a feature, too! Actually. I don't understand why this isn't already part of the framework - this has been bugging me since I started working with WPF 10 years ago.

I made many applications in WPF and I had run into this issue each and every time - there is nearly always some multithreading involved when creating real-world applications.

Each language and technology is good for certain use cases and not so good for others. WPF, just like many other Microsoft technologies, seems to be good only for writing tutorials and fails miserably when it comes to actual software development.
